In [1]:
import numpy as np

In [2]:
from math import factorial

### Задача 1

#### Контрольная работа состоит из пяти вопросов. На каждый вопрос приведено четыре варианта ответа, один из которых правильный.
#### Составьте закон распределения числа правильных ответов при простом угадывании. Найдите M(X), D(X).

#### Решение:

#### Пусть X - дискретная случайная величина, равная числу правильных ответов.
#### Она распределена по биноминальному закону с параметрами:

#### n = 5 (число вопросов), p = 1 / 4 (вероятность угадать ответ).

#### Поэтому вероятности вычисляем по формуле Бернулли:

#### P(X = k) = n! * (p ** k) * ((1 - p) ** (n - k))  / (k! * (n - k)!)

#### Получаем: 

In [3]:
n = 5
p = 1 / 4

In [4]:
x5 = k = 5
p5 = np.math.factorial(n) * (p ** k) * ((1 - p) ** (n - k)) / (np.math.factorial(k) * np.math.factorial(n - k))
p5

0.0009765625

In [5]:
x4 = k = 4
p4 = np.math.factorial(n) * (p ** k) * ((1 - p) ** (n - k)) / (np.math.factorial(k) * np.math.factorial(n - k))
p4

0.0146484375

In [6]:
x3 = k = 3
p3 = np.math.factorial(n) * (p ** k) * ((1 - p) ** (n - k)) / (np.math.factorial(k) * np.math.factorial(n - k))
p3

0.087890625

In [7]:
x2 = k = 2
p2 = np.math.factorial(n) * (p ** k) * ((1 - p) ** (n - k)) / (np.math.factorial(k) * np.math.factorial(n - k))
p2

0.263671875

In [8]:
x1 = k = 1
p1 = np.math.factorial(n) * (p ** k) * ((1 - p) ** (n - k)) / (np.math.factorial(k) * np.math.factorial(n - k))
p1

0.3955078125

In [9]:
x0 = k = 0
p0 = np.math.factorial(n) * (p ** k) * ((1 - p) ** (n - k)) / (np.math.factorial(k) * np.math.factorial(n - k))
p0

0.2373046875

#### Получили следующий закон распределения:

#### xi  0            1            2           3           4            5
#### pi  0.2373046875 0.3955078125 0.263671875 0.087890625 0.0146484375 0.0009765625

#### По формулам для биноминального распределения найдем M(X) = n * p и D(X) = n * p * (1 - p)

#### Получаем:

In [10]:
M = n * p
M

1.25

In [11]:
D = n * p * (1 - p)
D

0.9375

### Задача 2

#### Вероятность попадания в цель при одном выстреле равна 0.01.
#### Какова вероятность того, что число попаданий при 200 выстрелах будет не менее 5 и не более 10?

#### Решение:

#### По условию задачи: 
#### Вероятность попадания в цель при одном выстреле равна p = 0.01. Производится n = 200 выстрелов. 
#### Найти вероятность того, что цель будет поражена в точности k раз (будет k попаданий), где 5 <= k <= 10

#### Вероятность того, что цель будет поражена в этих n выстрелах ровно k раз, выражается формулой Бернулли:

#### Pn(k)=Ckn * p ** k * (1−p) ** (n - k), где Ckn - число сочетаний из n по k

In [12]:
p = 0.01
n = 200

In [13]:
k = 5
C5 = 2535650040

In [14]:
p ** k

1.0000000000000002e-10

In [15]:
q = (1 - p)
q

0.99

In [16]:
q ** (n - k)

0.14088441290426768

In [17]:
P5 = 2535650040 * 1.0000000000000002e-10 * 0.14088441290426768
P5

0.03572335672160829

In [18]:
k = 10
C10 = 22451004309013280

In [19]:
p ** k

1.0000000000000002e-20

In [20]:
q ** (n - k)

0.14814499154757946

In [21]:
P10 = 22451004309013280 * 1.0000000000000002e-20 * 0.14814499154757946
P10

3.326003843593443e-05

#### То есть вероятность того, что число попаданий при 200 выстрелах будет не менее 5 и не более 10,
#### находится в интервале от 0.03572335672160829 до 3.326003843593443e-05 = 0

### Задача 3

#### В условиях предыдущей задачи сколько выстрелов нужно сделать чтобы быть уверенным с вероятностью 0.9,
#### что хотя бы 10 раз будет совершено попадание?
#### Подсказка. Как посчитать P(X >= k) для величины X, имеющей распределение Пуассона? Здесь предстоит немножко покодить.

#### Решение:

#### Если проводится большое количество испытаний  n  и при этом вероятность  p  появления события  A  в отдельном испытании мала,
#### применяют формулу Пуассона для вычисления вероятности того, что событие произойдет  m  раз в  n  испытаниях:

#### Pm ≈ (λ ** m) * ((e ** (−λ)) / m!, где λ = n * p

#### Применяя формулу Пуассона, зная что m = 10, p = 0.01, Pm = 0.9, необходимо найти n.

#### Проведем пробные расчеты для различных n:

In [22]:
n = 1000
m = 10
p = 0.01
lambda_ = n * p

Pm = (lambda_** m / np.math.factorial(m)) * np.exp(-lambda_)
Pm

0.1251100357211333

In [23]:
n = 2000
m = 10
p = 0.01
lambda_ = n * p

Pm = (lambda_** m / np.math.factorial(m)) * np.exp(-lambda_)
Pm

0.005816306518345137

In [24]:
n = 800
m = 10
p = 0.01
lambda_ = n * p

Pm = (lambda_** m / np.math.factorial(m)) * np.exp(-lambda_)
Pm

0.09926153383153559

#### Может в задаче ошибка)) и Pm = 0.09?

In [25]:
m = 10
p = 0.01

for n in range(750, 1000):
    lambda_ = n * p
    Pm = (lambda_** m / np.math.factorial(m)) * np.exp(-lambda_)
    print("n = ", n, ", " "Pm = ", Pm)
    if Pm >= 0.09:
        print("Ответ: n = ", n)
        break
    else: 
        n += 5

n =  750 , Pm =  0.08583037040867353
n =  751 , Pm =  0.08611618420817457
n =  752 , Pm =  0.08640141781093084
n =  753 , Pm =  0.08668606324891183
n =  754 , Pm =  0.08697011258305054
n =  755 , Pm =  0.08725355790360638
n =  756 , Pm =  0.08753639133052492
n =  757 , Pm =  0.08781860501379449
n =  758 , Pm =  0.08810019113379965
n =  759 , Pm =  0.08838114190167139
n =  760 , Pm =  0.08866144955963406
n =  761 , Pm =  0.08894110638134928
n =  762 , Pm =  0.08922010467225631
n =  763 , Pm =  0.08949843676990946
n =  764 , Pm =  0.08977609504431222
n =  765 , Pm =  0.09005307189824763
Ответ: n =  765
